cpp/include/cudf/io/text/detail/tile_state.hpp

```diff
@@ -0,0 +1,144 @@
+
+#pragma once
+
+#include <cub/block/block_scan.cuh>
+
+#include <cuda/atomic>
+
+namespace cudf {
+namespace io {
+namespace text {
+namespace detail {
+
+enum class scan_tile_status : uint8_t {
+  oob,
+  invalid,
+  partial,
+  inclusive,
+};
+
+template <typename T>
+struct scan_tile_state_view {
+  uint64_t num_tiles;
+  cuda::atomic<scan_tile_status, cuda::thread_scope_device>* tile_status;
+  T* tile_partial;
+  T* tile_inclusive;
+
+  __device__ inline void set_status(cudf::size_type tile_idx, scan_tile_status status)
+  {
+    auto const offset = (tile_idx + num_tiles) % num_tiles;
+    tile_status[offset].store(status);
+  }
+
+  __device__ inline void set_partial_prefix(cudf::size_type tile_idx, T value)
+  {
+    auto const offset = (tile_idx + num_tiles) % num_tiles;
+    cub::ThreadStore<cub::STORE_CG>(tile_partial + offset, value);
+    tile_status[offset].store(scan_tile_status::partial);
+  }
+
+  __device__ inline void set_inclusive_prefix(cudf::size_type tile_idx, T value)
+  {
+    auto const offset = (tile_idx + num_tiles) % num_tiles;
+    cub::ThreadStore<cub::STORE_CG>(tile_inclusive + offset, value);
+    tile_status[offset].store(scan_tile_status::inclusive);
+  }
+
+  __device__ inline T get_prefix(cudf::size_type tile_idx, scan_tile_status& status)
+  {
+    auto const offset = (tile_idx + num_tiles) % num_tiles;
+
+    while ((status = tile_status[offset].load(cuda::memory_order_relaxed)) ==
+           scan_tile_status::invalid) {}
+
+    if (status == scan_tile_status::partial) {
+      return cub::ThreadLoad<cub::LOAD_CG>(tile_partial + offset);
+    } else {
+      return cub::ThreadLoad<cub::LOAD_CG>(tile_inclusive + offset);
+    }
+  }
+};
+
+template <typename T>
+struct scan_tile_state {
+  rmm::device_uvector<cuda::atomic<scan_tile_status, cuda::thread_scope_device>> tile_status;
```

Comment:
I'm surprised this compiles. The fact that it does seems like a bug. `cuda::atomic` isn't copyable nor moveable.

Reply:
after discussing we decided this use case isn't a bug since no move/copy takes place. However, `rmm::device_uvector` needs to be updated to prevent instantiation of any functions that require a move/copy of the value_type for types which cannot be moved/copied.